### ISSUE TITLE
Pier/landcover sorting

### ISSUE THREAD

**Issue by mrwojo** [gravitystorm/openstreetmap-carto#330]
Piers render over landcover. For natural=beach, this is generally good. For some other tags, such as tourism=attraction, leisure=park, and amenity=parking, this is often not good. layer=\* can't fix this problem because piers and landcovers are rendered separately.

[Brighton Pier](http://www.openstreetmap.org/#map=17/50.81574/-0.13766) and [Wildwood piers](http://www.openstreetmap.org/#map=18/38.98550/-74.80927) have attractions that should be pink areas on the pier.

[Santa Cruz Wharf](http://www.openstreetmap.org/#map=17/36.95971/-122.02147): A large part of it is a parking lot that's invisible except for a few bits that stick out from underneath the [pier area](http://www.openstreetmap.org/way/220162299).

Now it get's a little trickier:

[Santa Monica Pier](http://www.openstreetmap.org/#map=18/34.00874/-118.49732) should probably continue rendering over (or distinct from) the park underneath. But [how about this pier park](http://www.openstreetmap.org/#map=19/47.60685/-122.34197)? It should be green. Mappers may need to use layer=\* to differentiate these.

And importantly the render order is (from back to front):
1. landcover
2. water (non-ocean)
3. water-features (including piers)

With all that said we'd still need to have piers render over water and water over landcover.

Any ideas on how to break the circle or otherwise improve the existing the behavior?

**Comment by jeisenbe** [comment]
Now that we have switched to water polygons, and landcover is rendered under water-areas and piers are rendered above water-areas, I don't think there is a way to solve this issue. We also removed landcover sorting by layer and #53 was closed.

So we can close this issue as wontfix?

**Comment by jeisenbe** [comment]
Actually there is one possible solution, but it involves more `comp-op` and an extra layer:

0) Implement #3854, which would mean that water areas would be "cut out" of the landcover using `dst-out`
1) Change order to first "cut out" oceans and other water areas prior to rendering landcover.
2) Next render piers in `land-color` (or any other color),
3) Then render landcover with `comp-op: dst-in`, which means that landcover would only be rendered on top of land and piers
4) After that, water areas would be rendered behind as described in #3854

**Comment by controversial** [comment]
Is this blocked on #3854? Or could a similar approach to #5067 be started without waiting for any blockers here?

**Comment by controversial** [comment]
Got it, thanks!

There was a discussion in the OSM US slack over the weekend about the piers in NYC—NYC has dozens of piers that are developed with mixed use as athletic facilities, public parks, etc.

In NYC, unfortunately, several mappers have *removed* `man_made=pier` from many piers—simply because otherwise much of their mapping effort does not appear on Carto. Compare piers _without_ `man_made=pier` (incorrect) vs with `man_made=pier` (correct tagging, but no landcover rendering):

<img width="400" height="auto" alt="Piers without man_made=pier" src="https://github.com/user-attachments/assets/feb89b7e-3db8-4400-be26-4a8dfa67a084" /> <img width="400" height="auto" alt="Pier tagged with man_made=pier" src="https://github.com/user-attachments/assets/e8b2edd9-3625-4eff-85bf-46622145371b" />

---

This is just to give one example of how the pier sorting issue hurts the goal “to serve as feedback for mappers and encourage correct mapping.”

**Comment by controversial** [comment]
If there’s not a “natural” order for landcover vs piers (`natural=beach` should usually go below, but not in [all cases]()) I wonder if it’s possible to do size-based sorting as @matthijsmelissen suggested? 

Sorting by area would be the simpler approach, although maybe a more complex operation to find landcover polygons that are contained completely within a `man_made=pier` would be the most accurate?

Still, for cases like “parking lots under piers” the only unambiguous way for mappers to specify the order would be to use `layer`

**Comment by imagico** [comment]
In principle you could do something like that.  But it won't work in a number of cases:

* features under a pier that are fully covered by the pier
* piers mapped with linear ways
* piers split in mapping (because of different secondary tags or simply for convenience of mapping)
* features that extend beyond the pier onto other things (like a pitch/park/whatever that is partly on the pier, partly on solid ground).

Or in other words: The use of `ST_Contains()` would be as a means to better *guess* if a landcover feature is on or below a pier.  But is is clearly not a reliable indicator for that.

What we really need to consider is IMO (a) if there is a consensus method among mappers to record the information that a certain feature is located on top of a pier rather than below it and (b) if and how we can use that information to adjust rendering.  And if (a) is not the case then maybe it is a good idea for us not to try guessing - because then important information to non-ambiguously interpret the data is clearly missing.

**Comment by controversial** [comment]
I think making the *guess* that “features contained within a pier’s area are built atop of the pier” (as opposed to under, which seems like the _much_ rarer case to me) would be a *huge* improvement over the status quo, because it would be correct in many more cases than the current behavior!

Your point stands that more information is needed, in many cases, to interpret the map data truly unambiguously. I agree with your assessment that a hypothetical tag like `location=pier` would make the features’ relationships clear enough to be rendered confidently; unfortunately i’m not aware of any such usage in the wild.

Under **current** tagging practices/consensus, I think it would be more appropriate to resolve an ambiguity of the form “is this parking lot below the pier” by tagging the _parking lot_ with the well-established [`location=surface`](https://wiki.openstreetmap.org/wiki/Tag:location%3Dsurface) (“*Feature is on the surface of the ground; useful where it is not obvious that object is on the ground*”). `location=surface` would allow a mapper, within the current “consensus” tagging schema, to resolve the ambiguity presented by the (reasonable, imo) assumption that a parking lot within a pier is built on the pier.

---

I disagree that leaving the behavior as-is allows us “not to try guessing”; my perspective is that the current behavior is _already_ to make a “guess” that:
- buildings, roads, trees, etc. are always _on top of_ a pier
- landcover is always below it (despite the fact that tags don’t contain the information required to know this unambiguously).

In my opinion, absent your (a), developing a heuristic to “guess right” in many _more_ cases than we currently do would make a substantial improvement to the map, compared to the alternative (never doing anything?)

**Comment by controversial** [comment]
Actually, to the question of 
> [is there] a consensus method among mappers to record the information that a certain feature is located on top of a pier rather than below it

taginfo [shows](https://taginfo.openstreetmap.org/tags/man_made=pier#combinations) that there are 8369 uses of `layer=*` on `man_made=pier`; anecdotally, this also the solution volunteered by several members of the OSM US slack when I posed your question about the consensus method!

The [documented](https://wiki.openstreetmap.org/wiki/Key:layer) purpose of `layer`:
> to describe vertical relationships between crossing or overlapping features

seems to me like a natural fit for this issue of how to describe vertical relationships between overlapping features.

so, I would argue `layer` _is_ the consensus method to record the information that a certain features is located on top of or below a pier.

---

then I’d propose we could feasibly create a `pier_landcover` layer above `piers-poly` by some query of the rough form:

```sql
SELECT ST_Intersection(landcover.way, pier.way) AS way, ...
FROM planet_osm_polygon AS landcover
JOIN planet_osm_polygon AS pier
  ON pier.man_made = 'pier'
 AND is_landcover(landcover)
 AND ST_Intersects(pier.way, landcover.way)
 AND landcover.layer > pier.layer
```

(though I do still believe that “defaulting” landcover that is _within_ piers to appear _on top of the pier_ is a better default—absent layering information from tags—than the current guess)

**Comment by controversial** [comment]
I take your point about making any heuristic used here one that would be intuitively understandable for mappers, and reproducible by other maps. I hadn’t considered those goals, and I definitely see where you’re coming from there. Thanks for explaining!

I also agree that it’s not feasible to render all landcover without `location=surface` after piers (this would break the fairly common use case of drawing a pier over a beach). I had suggested `location=surface` mostly as usable in conjunction with an intersection or containment-based heuristic, which it does seem reasonable that we would lean away from on the basis of being intuitive for mappers.

---

I’m still not sure I follow your argument that `layer` isn’t a good fit.

> The main thing to realize about layer=* is that it is purely local and purely relative w.r.t. intersecting or closeby but unconnected features ... the information needed is not so much a relative location

it seems to me like the information we need here is _exactly_ the relative location of the pier vs the landcover! e.g. if a mapper draws _overlapping_ features:
- `amenity=parking` with `layer=1`
- `man_made=pier` with `layer=2`
- `leisure=pitch` with `layer=3`

it would appear to me like:
1. this would be _exactly_ the same ordering behavior that’s “well defined ... with roads,” i.e. purely local and purely relative.
2. this _unambiguously_ provides the relative ordering information about which landcover is on the earth surface and which is above the pier.
3. this heuristic is easy for mappers to understand, because it follows what’s used on roads.

I see your point that `location` would be semantically richer than `layer`, but currently “in the wild” there appears to be at least _some_ usage of `layer` for piers and _some_ consensus that layer is appropriate for this use case, but virtually _no_ instances of `location=pier`.

It‘s my understanding that Carto wouldn’t actually want to make any rendering changes based on tagging practices like `location=pier` unless/until such time as there‘s documented community usage + consensus about this tagging practice, so deciding to prefer `location=pier` for disambiguating this case might mean, in practice, that the issue never gets solved!

**Comment by imagico** [comment]
The semantics of `layer=*` w.r.t. linear roads are discussed in detail [here](https://imagico.de/blog/en/navigating-the-maze-part-1/) - the OSM wiki does not provide very useful information here on the de facto practice because the key aspects to keep in mind (that `layer=*` documents the relative vertical positions of intersecting and closeby roads that do not connect) are fairly hidden in the text and not easy to find for readers.  This semantic principle is very specific (it only applies to linear roads, it already fails on road polygons) but is also fairly consistently applied by mappers.

The main part of cargo cult tagging with layer=* for non-road features are attempts of mappers to hand draw the map, i.e. not to document verifiable information about the geographic reality but to try to tell the map designers ad hoc, on a case-by-case basis, how to draw their maps.

But again: In OSM-Carto we are trying to be agnostic w.r.t. decisions of mappers what tags to use.  If there is consensus among mappers to overload `layer=*` in a different meaning for landcover polygons and `man_made=pier` and it is used in that context in a consistent fashion that does not interfere with the use on roads (which does not really appear to be possible to me - but i have not thought about it much) then we would try to interpret that of course.

> It‘s my understanding that Carto wouldn’t actually want to make any rendering changes based on tagging practices like location=pier unless/until such time as there‘s documented community usage + consensus about this tagging practice, so deciding to prefer location=pier for disambiguating this case might mean, in practice, that the issue never gets solved!

Proper documentation of the mapping practice is important but we ultimately base our rendering decisions on how mappers de facto map.  `location=*` here has a bit of a starting advantage because it is already used widely to indicate the vertical location of features relative to ground level, for example to indicate something is located on the roof of a building (which for map rendering is a problem very similar to the piers).  Among the [currently used values](https://taginfo.openstreetmap.org/keys/location#values) [aboveground](https://taginfo.openstreetmap.org/tags/location=aboveground) would probably be the closest match, but `location=pier` would make sense as well in analogy to `location=roof` and `location=bridge` of course.

For `layer=*` mappers would need to consider the semantics in case this is applied to landcover and `man_made=pier`.  Questions you would need to consider for example are

* What real world meaning does it have if a `natural=wood` polygon with `layer=1` overlaps a `natural=water` polygon with `layer=0`.  Is that a floating forest?  Does this meaning change if there is a third feature involved with a layer tag (or with no layer tag)?
* How does this use of `layer=*` interact with the established use on roads?  What does it mean in a road with `layer=-1` intersects a `landuse=farmland` polygon with `layer=0`.  Is that a tunnel despite there being no `tunnel=*` tag?  Same if the farmland is `layer=1` and the road has no layer tag? What does it mean when a `highway=*` way with `layer=1` (and typically `bridge=yes`) shares an end node with a linear `man_made=pier` with `layer=0` or no layer (would be a fairly typical situation of a floating pier connected to land via a gangway)?

Note these question are nothing to discuss here - but they are questions that would be advisable to be discussed by mappers and to achieve consensus on before starting to use `layer=*` on landcover and pier features.

In terms of chances for getting a change into the style: Developers would probably have a relatively low motivation and we would be also hesitant to implement a complex change just for a fairly niche use case like micromapping piers.  But if the same design concept would work for piers, bridges and building roofs (no matter if all of them use the same tagging or not) then that is a pretty broad set of issues that would be solved - and the chances for getting this into OSM-Carto would be good (provided, of course, that we have a consistent mapping practice).
